RHOAIENG-27813: extend coverage for the runtime config map tests

jstourac opened the pull request:
https://issues.redhat.com/browse/RHOAIENG-27813 (https://issues.redhat.com/browse/RHOAIENG-24545)

## Description
<!--- Describe your changes in detail -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Tests**
  - Refactored and expanded test coverage for ConfigMap mounting in Notebooks, including scenarios for empty ConfigMaps and ImageStream-driven ConfigMap creation.
  - Improved test structure, clarity, and type safety.
  - Added explicit checks for ConfigMap presence and volume mounts in Notebooks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -34,107 +37,188 @@ var _ = Describe("When Creating a notebook should mount the configMap", func() {
-						"data": map[string]string{
-							"datascience.json": `{"image_name":"quay.io/opendatahub/test"}`,
-						},
+				ConfigMap: &corev1.ConfigMap{
```
I'd not mind separate PR for this, but it should be done one way or another

jstourac replied:
Nope, I completely agree, that is a good point. Updated with the latest push.

jstourac commented:
@jiridanek thank you very much for your review. I have addressed your concerns. Also, I stabilized the test hopefully and also rebased against the main branch.

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -34,107 +37,188 @@ var _ = Describe("When Creating a notebook should mount the configMap", func() {
+					Eventually(func(g Gomega) {
+						err := cli.Get(ctx, client.ObjectKey{Name: RuntimeImagesCMName, Namespace: Namespace}, configMap)
+						g.Expect(err).ToNot(HaveOccurred())
+					}, 15*time.Second, time.Second).Should(Succeed())
```
fine with me

the reason for suggesting semantic timeouts like `RESOURCE_RECONCILIATION_TIMEOUT` is that one can take a look at definition and think something like "I believe that having 60 second reconciliation timeout is suspicious, maybe controller is not reconciling as fast as it should..."

or if we even shift the tests to run at kind, maybe it will justify doubling some timeouts and not other, and if we have them named according to purpose, the decision may be easier...

jstourac replied:
updated to have these separated

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -34,107 +35,196 @@ var _ = Describe("When Creating a notebook should mount the configMap", func() {
 			notebookName      string
-			ConfigMap         *unstructured.Unstructured
+			ConfigMap         *corev1.ConfigMap
+			imageStream       *imagev1.ImageStream
```
maybe a comment that either imagestream or configmap is to be provided? didn't you say you want to remove specifying configmap?

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -7,10 +7,10 @@ import (
 	nbv1 "github.com/kubeflow/kubeflow/components/notebook-controller/api/v1"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	imagev1 "github.com/openshift/api/image/v1"
 	corev1 "k8s.io/api/core/v1"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"sigs.k8s.io/controller-runtime/pkg/client"
```
could you group the imports according to meaning?

leave blank line between the `. ` imports, then do the schema imports like `nbv1`, `imagev1`, and then another blank line and then the other stuff?

jiridanek replied:
in go it usually remains how you leave it; afaik there has not been any strong fixation on tooling for import ordering in go

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -34,107 +35,196 @@ var _ = Describe("When Creating a notebook should mount the configMap", func() {
+					}
+
+					// Check volumeMounts
+					By("Fetching the created Notebook CR as typed object and volumeMounts check")
+					typedNotebook := &nbv1.Notebook{}
+					Eventually(func(g Gomega) {
+						err := cli.Get(ctx, client.ObjectKey{Name: testCase.notebookName, Namespace: Namespace}, typedNotebook)
+						g.Expect(err).ToNot(HaveOccurred())
+
+						c := typedNotebook.Spec.Template.Spec.Containers[0]
+
+						foundMount := false
+						for _, vm := range c.VolumeMounts {
+							if vm.Name == testCase.expectedMountName && vm.MountPath == testCase.expectedMountPath {
+								foundMount = true
+							}
+						}
+
+						if testCase.expectedMountName != "" {
+							g.Expect(foundMount).To(BeTrue(), "expected VolumeMount not found")
+						} else {
+							g.Expect(foundMount).To(BeFalse(), "unexpected VolumeMount found")
+						}
+					}, timeout, interval).Should(Succeed())
+
+					// Check volumes
+					foundVolume := false
+					for _, v := range typedNotebook.Spec.Template.Spec.Volumes {
+						if v.Name == testCase.expectedMountName && v.ConfigMap != nil && v.ConfigMap.Name == RuntimeImagesCMName {
+							foundVolume = true
+						}
+					}
+					if testCase.expectedMountName != "" {
+						Expect(foundVolume).To(BeTrue(), "expected ConfigMap volume not found")
+					} else {
+						Expect(foundVolume).To(BeFalse(), "unexpected ConfigMap volume found")
```
It did screw up the Eventually, it has to refetch notebook, but other than that, good.

jiridanek replied:
@coderabbitai we won't be doing it in this pr, please create a detailed github issue to track the work

jiridanek replied:
* https://github.com/opendatahub-io/kubeflow/issues/635

jiridanek commented:
> RHOAIENG-24545: modified the runtime config map tests #623

imperative mood in commit messages? ;p

jstourac commented on components/odh-notebook-controller/controllers/suite_test.go:
```diff
@@ -74,7 +74,7 @@ var (
 
 const (
-	timeout                            = time.Second * 10
+	timeout                            = time.Second * 30
```
I put it back to the original value.

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -34,107 +35,196 @@ var _ = Describe("When Creating a notebook should mount the configMap", func() {
+					}, timeout, interval).Should(Succeed())
+
+					if testCase.ConfigMap == nil {
+						// Let's check the content of the ConfigMap created from the given ImageStream.
```
@coderabbitio we will address this as part of https://github.com/opendatahub-io/kubeflow/issues/634. make a note of it.

jiridanek commented:
@coderabbitai please remember that you are always supposed to open github issues on the opendatahub-io/kubeflow repository unless instructed otherwise

jiridanek commented:
@coderabbitai I apologize, wrong link, you did it wrong with respect to https://github.com/kubeflow/kubeflow/issues/7730

jiridanek commented:
> │ Help me learn! │

![image](https://github.com/user-attachments/assets/2bfa43e5-d2a4-4c5a-86a7-3a8825ae4272)

![image](https://github.com/user-attachments/assets/57f4ecdc-9f52-4889-bf6d-aa7a87e198d0)

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
https://blog.thecodewhisperer.com/permalink/modularity-details-pick-one

jiridanek replied:
@coderabbitai reflect on the DAMP idea when writing test code

jiridanek commented:
@coderabbitai full

jiridanek commented on components/odh-notebook-controller/controllers/notebook_runtime_test.go:
```diff
@@ -3,25 +3,132 @@ package controllers
+					_ = cli.Delete(ctx, notebook, &client.DeleteOptions{})
+					_ = cli.Delete(ctx, testCase.ConfigMap, &client.DeleteOptions{})
+
+					// wait until deleted
+					By("Waiting for the Notebook to be deleted")
```
the `By` serves by way of the comment; why have both a comment and a By that says exactly the same thing?
